src/compiler/transformers/esnext.ts

```diff
@@ -101,11 +114,204 @@ namespace ts {
+            const statements = createPrivateNameWeakMapDeclarations(
+                currentPrivateNameEnvironment()
+            );
+            if (statements.length) {
+                node = updateClassDeclaration(
+                    node,
+                    node.decorators,
+                    node.modifiers,
+                    node.name,
+                    node.typeParameters,
+                    node.heritageClauses,
+                    transformClassMembers(node.members)
+                );
+            }
+            statements.unshift(node);
+            endPrivateNameEnvironment();
+            return statements;
+        }
+
+        function visitClassExpression(node: ClassExpression) {
+            startPrivateNameEnvironment();
+            node = visitEachChild(node, visitorCollectPrivateNames, context);
+            node = visitEachChild(node, visitor, context);
+            const expressions = createPrivateNameWeakMapAssignments(
+                currentPrivateNameEnvironment()
+            );
+            if (expressions.length) {
+                node = updateClassExpression(
+                    node,
+                    node.modifiers,
+                    node.name,
+                    node.typeParameters,
+                    node.heritageClauses,
+                    transformClassMembers(node.members)
+                );
+            }
+            expressions.push(node);
+            endPrivateNameEnvironment();
+            return expressions.length > 1 ? createCommaList(expressions) : expressions[0];
+        }
+
+        function startPrivateNameEnvironment() {
+            // Create private name environment.
+            privateNameEnvironmentStack[++privateNameEnvironmentIndex] = {};
+            return currentPrivateNameEnvironment();
+        }
+
+        function endPrivateNameEnvironment(): PrivateNameEnvironment {
+            const privateNameEnvironment = currentPrivateNameEnvironment();
+            // Destroy private name environment.
+            delete privateNameEnvironmentStack[privateNameEnvironmentIndex--];
+            return privateNameEnvironment;
+        }
+
+        function createPrivateNameWeakMapDeclarations(environment: PrivateNameEnvironment): Statement[] {
+            return Object.keys(environment).map(name => {
+                const privateName = environment[name];
+                return createVariableStatement(
+                    /* modifiers */ undefined,
+                    [createVariableDeclaration(privateName.weakMap,
+                                               /* typeNode */ undefined,
+                        createNew(
+                            createIdentifier("WeakMap"),
+                                                   /* typeArguments */ undefined,
+                                                   /* argumentsArray */ undefined
+                        ))]
+                );
+            });
+        }
+

```

Comment:
`createPrivatenameWeakMapAssignments` generates something like the following when targeting ES5 and below. Why is the _foo declared above, then assigned in a comma expression?

```
var _classPrivateFieldSet = function (receiver, privateMap, value) { if (!privateMap.has(receiver)) { throw new TypeError("attempted to set private field on non-instance"); } privateMap.set(receiver, value); return value; };
var _foo;
"use strict";
var x = (_foo = new WeakMap, /** @class */ (function () {
    function A() {
        _foo.set(this, void 0);
        _classPrivateFieldSet(this, _foo, "foo");
    }
    return A;
}()));
```

Reply:
Wait, I think I see. My example is bad. In other cases, you really want the internals of the class to close over a different WeakMap each time the class is evaluated.

Reply:
Better example:


```js
var _classPrivateFieldSet = function (receiver, privateMap, value) { if (!privateMap.has(receiver)) { throw new TypeError("attempted to set private field on non-instance"); } privateMap.set(receiver, value); return value; };
var _foo;
"use strict";
var xs = [];
for (var i = 0; i < 10; i++) {
    xs.push((_foo = new WeakMap, /** @class */ (function () {
        function X() {
            _foo.set(this, void 0);
            _classPrivateFieldSet(this, _foo, "foo");
        }
        return X;
    }())));
}
```